File: learning-core/executable/queue/cpp/queue.cpp
#include <deque>
#include <stdexcept>
#include <string>

class queue_full : public std::runtime_error {
public:
    using std::runtime_error::runtime_error;
};

class BoundedQueue {
public:
    explicit BoundedQueue(std::size_t capacity) : capacity_(capacity) {}

    void put(std::string item) {
        if (items_.size() >= capacity_) {
            throw queue_full("queue capacity reached");
        }
        items_.push_back(std::move(item));
    }

    std::string get() {
        auto item = std::move(items_.front());
        items_.pop_front();
        return item;
    }

    std::size_t size() const { return items_.size(); }

private:
    std::size_t capacity_;
    std::deque<std::string> items_;
};

File: learning-core/executable/queue/go/queue.go
package queue

import "errors"

var ErrFull = errors.New("queue capacity reached")

type BoundedQueue struct {
	items    []string
	capacity int
}

func New(capacity int) *BoundedQueue {
	return &BoundedQueue{capacity: capacity}
}

func (q *BoundedQueue) Put(item string) error {
	if len(q.items) >= q.capacity {
		return ErrFull
	}
	q.items = append(q.items, item)
	return nil
}

func (q *BoundedQueue) Get() (string, bool) {
	if len(q.items) == 0 {
		return "", false
	}
	item := q.items[0]
	q.items = q.items[1:]
	return item, true
}

func (q *BoundedQueue) Len() int { return len(q.items) }
